### PR TITLE
feat(models): reference context for relative paths

### LINKED ISSUE #452: cli: fix paths of make export
**Describe the bug**

When using the `make` export option of `renku` as in:

```
renku log --format=Makefile outputs/plot.pdf
```
paths are off:
```
outputs/plot.pdf: src/plot.py outputs/simdata.csv
        @python ../../src/plot.py --input ../../outputs/simdata.csv
outputs/simdata.csv: src/simulate.R
        @Rscript ../../src/simulate.R outputs/simdata.csv
```
**Expected behavior**

Output should be:

```
outputs/plot.pdf: src/plot.py outputs/simdata.csv
        @python src/plot.py --input outputs/simdata.csv
outputs/simdata.csv: src/simulate.R
        @Rscript  src/simulate.R outputs/simdata.csv
```
Unfortunately, even fixing that I have problems at run time (see attached example):

```bash
sysadmins-MacBook-Pro:mm cchoirat$ make
Traceback (most recent call last):
  File "src/plot.py", line 7, in <module>
    sim_data.plot()
  File "/usr/local/anaconda3/lib/python3.7/site-packages/pandas/plotting/_core.py", line 2939, in __call__
    sort_columns=sort_columns, **kwds)
  File "/usr/local/anaconda3/lib/python3.7/site-packages/pandas/plotting/_core.py", line 1970, in plot_frame
    **kwds)
  File "/usr/local/anaconda3/lib/python3.7/site-packages/pandas/plotting/_core.py", line 1798, in _plot
    plot_obj.generate()
  File "/usr/local/anaconda3/lib/python3.7/site-packages/pandas/plotting/_core.py", line 249, in generate
    self._compute_plot_data()
  File "/usr/local/anaconda3/lib/python3.7/site-packages/pandas/plotting/_core.py", line 364, in _compute_plot_data
    'plot'.format(numeric_data.__class__.__name__))
TypeError: Empty 'DataFrame': no numeric data to plot
```
(I also believe that spaces before `@` should be a tab.)

[minimake.zip](https://github.com/SwissDataScienceCenter/renku-python/files/2945901/minimake.zip)

### PR DESCRIPTION
**Pull request by jirikuncar**
(closes #452)

### CODE REVIEW

File: renku/models/_jsonld.py [hunk 1]
```diff
@@ -333,6 +335,19 @@ def from_jsonld(cls, data):
                for k, v in compacted.items() if k in fields}
         )
 
+    @classmethod
+    def from_yaml(cls, path):
+        """Return an instance from a YAML file."""
+        import yaml
+        # import functools
+        # cls = functools.partial(cls, __reference__=file_)
```
Comment (jsam) [inline renku/models/_jsonld.py]: Can we clean that up or put TODO mark?

File: renku/models/cwl/_ascwl.py [hunk 2]
```diff
@@ -40,16 +40,34 @@ def __init__(cls, name, bases, namespace):
         cls.registry[name] = cls
 
 
-class CWLClass(object, metaclass=CWLType):
+# @attr.s
+class CWLClass(metaclass=CWLType):
     """Include ``class`` field in serialized object."""
 
+    # __reference__ = attr.ib(default=None, kw_only=True)
```
Comment (jsam) [inline renku/models/cwl/_ascwl.py]: Cleanup?

File: renku/models/cwl/_ascwl.py [hunk 3]
```diff
@@ -40,16 +40,34 @@ def __init__(cls, name, bases, namespace):
...
         class_name = data.get('class', None)
         cls = cls.registry.get(class_name, cls)
         return cls(**{k: v for k, v in iteritems(data) if k != 'class'})
 
+    @classmethod
+    def from_yaml(cls, path):
+        """Return an instance from a YAML file."""
+        import yaml
+        # import functools
+        # cls = functools.partial(cls, __reference__=file_)
```
Comment (jsam) [inline renku/models/cwl/_ascwl.py]: Cleanup?

File: renku/models/cwl/_ascwl.py [hunk 4]
```diff
@@ -40,16 +40,34 @@ def __init__(cls, name, bases, namespace):
...
+
+        with path.open(mode='r') as fp:
+            # return cls.from_cwl(yaml.load(fp))
```
Comment (jsam) [inline renku/models/cwl/_ascwl.py]: Cleanup?

File: renku/api/datasets.py [hunk 5]
```diff
@@ -80,7 +79,7 @@ def with_dataset(self, name=None):
                 if path.exists():
                     with path.open('r') as f:
                         source = yaml.load(f) or {}
-                    dataset = Dataset.from_jsonld(source)
+                    dataset = Dataset.from_jsonld(source, __reference__=path)
```
Comment (jsam) [inline renku/api/datasets.py]: This line builds an incomplete model tree where each ```DatasetFile.dataset = None```. Do you have any ideas on how we could easily fix this without additional iteration and setting it manually?

Comment (jirikuncar) [reply renku/api/datasets.py]: we won't need the `dataset` property anymore

File: renku/models/provenance/activities.py [hunk 6]
```diff
@@ -418,12 +417,12 @@ def _load(step):
                 return step.run
 
             if self.commit:
+                import yaml
```
Comment (jsam) [inline renku/models/provenance/activities.py]: Is this imported second time or I'm blind?

Comment (jirikuncar) [reply renku/models/provenance/activities.py]: I have removed the global one.

**Pull request merged**